driver-core/src/main/com/mongodb/internal/TimeoutContext.java

```diff
@@ -200,30 +184,43 @@ public long getMaxAwaitTimeMS() {
         return timeoutSettings.getMaxAwaitTimeMS();
     }
 
-    public long getMaxTimeMS() {
-        return notNull("Should never be null", maxTimeSupplier.get());
-    }
+    public void runMaxTimeMSTimeout(final Runnable onInfinite, final LongConsumer onRemaining,
+            final Runnable onExpired) {
+        if (maxTimeSupplier != null) {
+            runWithFixedTimout(maxTimeSupplier.get(), onInfinite, onRemaining);
+            return;
+        }
 
-    private long calculateMaxTimeMS() {
-        long maxTimeMS = timeoutOrAlternative(timeoutSettings.getMaxTimeMS());
-        if (timeout == null || timeout.isInfinite()) {
-            return maxTimeMS;
+        if (timeout != null) {
+            timeout.shortenBy(minRoundTripTimeMS, MILLISECONDS)
+                    .run(MILLISECONDS, onInfinite, onRemaining, onExpired);
+        } else {
+            runWithFixedTimout(timeoutSettings.getMaxTimeMS(), onInfinite, onRemaining);
         }
-        if (minRoundTripTimeMS >= maxTimeMS) {
-            throw createMongoTimeoutException();
+    }
+
+    private static void runWithFixedTimout(final long ms, final Runnable onInfinite, final LongConsumer onRemaining) {
+        if (ms == 0) {
+            onInfinite.run();
+        } else {
+            onRemaining.accept(ms);
         }
-        return maxTimeMS - minRoundTripTimeMS;
     }
 
-    public void setMaxTimeSupplier(final MaxTimeSupplier maxTimeSupplier) {
-        this.maxTimeSupplier = maxTimeSupplier;
+    public void resetToDefaultMaxTime() {
+        this.maxTimeSupplier = null;
+    }
+
+    public void setMaxTimeOverride(final long maxTimeMS) {
```

Comment:
[Optional] Currently, this method accepts a pre-calculated maxTimeMS. One of the original intents behind considering a supplier was to enable operations to calculate maxTimeMS right before the command is sent, potentially reducing timing errors. In my local testing indicates that this delayed calculation method can indeed be more precise, reducing the error margin to about 1-3 milliseconds compared to pre-calculated values.

For now, there's no need for a method with delayed maxTimeMS calculation, but this could change. I suggest adding the Javadoc to reflect that it accepts pre-calculated values and to mention the future possibility of using a supplier lambda for more precise timing, ensuring we remember this option moving forward.